crates/fluvio-sc/src/controllers/mirroring/controller.rs

```diff
@@ -0,0 +1,220 @@
+use std::time::{Duration, SystemTime};
+use fluvio_controlplane_metadata::mirroring::{MirrorConnect, MirroringRemoteClusterRequest};
+
+use fluvio_socket::{ClientConfig, MultiplexerSocket};
+use futures_util::StreamExt;
+use fluvio_future::{task::spawn, timer::sleep};
+use fluvio_sc_schema::{
+    core::MetadataItem,
+    mirror::{ConnectionStatus, MirrorPairStatus, MirrorSpec, MirrorStatus, MirrorType},
+    mirroring::ObjectMirroringRequest,
+    topic::{MirrorConfig, RemoteMirrorConfig, ReplicaSpec, TopicSpec},
+    TryEncodableFrom,
+};
+use fluvio_stream_dispatcher::store::StoreContext;
+use tracing::{debug, error, info, instrument};
+use anyhow::{anyhow, Context, Result};
+
+use crate::core::SharedContext;
+
+const MIRRORING_CONTROLLER_INTERVAL: u64 = 5;
+const MIRRORING_CONTROLLER_RETRY_INTERVAL: u64 = 10;
+
+// This is the main controller for the mirroring feature.
+// Remote Clusters will connect to the home clusters.
+pub struct MirroringController<C: MetadataItem> {
+    mirrors: StoreContext<MirrorSpec, C>,
+    topics: StoreContext<TopicSpec, C>,
+}
+
+impl<C: MetadataItem> MirroringController<C> {
+    pub fn start(ctx: SharedContext<C>) {
+        let controller = Self {
+            mirrors: ctx.mirrors().clone(),
+            topics: ctx.topics().clone(),
+        };
+
+        info!("starting mirroring controller");
+        spawn(controller.dispatch_loop());
+    }
+
+    #[instrument(skip(self), name = "MirroringControllerLoop")]
+    async fn dispatch_loop(self) {
+        info!("started");
+        loop {
+            if let Err(err) = self.inner_loop().await {
+                error!("error with inner loop: {:#?}", err);
+                debug!(
+                    "sleeping {} seconds try again",
+                    MIRRORING_CONTROLLER_RETRY_INTERVAL
+                );
+                sleep(Duration::from_secs(MIRRORING_CONTROLLER_RETRY_INTERVAL)).await;
+            }
+        }
+    }
+
+    #[instrument(skip(self))]
+    async fn inner_loop(&self) -> Result<()> {
+        debug!("initializing listeners");
+
+        loop {
+            let mirrors_from_home = self
+                .mirrors
+                .store()
+                .read()
+                .await
+                .values()
+                .filter_map(|r| match r.spec().mirror_type.clone() {
+                    MirrorType::Home(h) => Some(h),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+
+            //TODO: we need to have a strategy to choose which home to connect
+            if let Some(mirror) = mirrors_from_home.first() {
+                //send to home cluster the connect request
+                //TODO: handle TLS
+
+                let home_config = ClientConfig::with_addr(mirror.public_endpoint.clone());
```

Comment:
This should be using profile downloaded from remote config otherwise TLS will not work